testsuite/tests/misc/weaklifetime.ml

```diff
@@ -48,11 +55,9 @@ let check_and_change i j =
     data.(i).objs.(j) <- Present x;
     Weak.set data.(i).wp j (Some x);
   | Present _, true ->
-    if Random.int 10 = 0 then begin
-      data.(i).objs.(j) <- Absent gc1;
-      let gc2 = gccount () in
-      if gc1 <> gc2 then data.(i).objs.(j) <- Absent gc2;
-    end
+    if Random.int 10 = 0 then (erase [@ocaml.tailcall]) i j gc1
+    (* Must be a tail-call to get out of the scope of the binding of
+       [data.(i).objs.(j)] that is introduced by the pattern-matching. *)
```

Comment:
If the aim is to ensure that the previous value of `data.(i).objs.(j)` is not retained on the bytecode stack, would the following suffice?

```ocaml
type change = No_change | Set_present | Maybe_set_absent

let check_and_change i j =
  let gc1 = gccount () in
  let change =
    (* we only read data.(i).objs.(j) in this local binding to ensure
        that it does not remain reachable on the bytecode stack
        in the rest of the function below, when we overwrite the value
        and try to observe its collection.  *)
    match data.(i).objs.(j), Weak.check data.(i).wp j with
    | Present x, false -> assert false
    | Absent n, true -> assert (gc1 <= n+1); No_change
    | Absent _, false -> Make_present
    | Present _, true ->
      if Random.int 10 = 0 then Make_absent else No_change
  in
  match change with
  | No_change -> ()
  | Make_present ->
      let x = Array.make (1 + Random.int 10) 42 in
      data.(i).objs.(j) <- Present x;
      Weak.set data.(i).wp j (Some x);
  | Make_absent ->
      data.(i).objs.(j) <- Absent gc1;
      let gc2 = gccount () in
      if gc1 <> gc2 then data.(i).objs.(j) <- Absent gc2;
```

Reply:
(The point, I guess, is that I find it more natural to use lexical scoping to scope lifetimes, rather than to use tail-recursion to break out of the lexical-scoping-related lifetime.)

Reply:
I agree with @gasche's point about scoping, but I'm also eager to have this test fixed, as it is ruining our CI right now.  Either @damiendoligez rewrites as suggested very soon, or the fix goes in unchanged.

Reply:
I like @gasche's version. I'll copy it and re-test with my scaffolding to make sure it works.